robot/control/Src/motion-control/RobotController.cpp

```diff
@@ -0,0 +1,186 @@
+#include "motion-control/RobotController.hpp"
+#include "rc-fshare/robot_model.hpp"
+#include "mtrain.hpp"
+#include "MicroPackets.hpp"
+#include <cmath>
+
+extern DebugInfo debugInfo;
+
+/**
+ * Edited sign function to always return -1 or 1
+ * 
+ * Returns 1 if val >= 0, -1 if val < 0
+ */
+template <typename T> int sgn(T val) {
+    return (T(0) <= val) - (val < T(0));
+}
+
+template <typename T> T bounded(T val, T min, T max) {
+    return (val < max && val > min);
+}
+
+double limit(double val, double maxabs) {
+    if (val > maxabs) {
+        return maxabs;
+    } else if (val < -maxabs) {
+        return -maxabs;
+    } else {
+        return val;
+    }
+}
+
+double JERK_LIMIT = 10.0;
+double TICK_TIME = 5e-3;
+double ROBOT_MASS_X = 6.35;
+double ROBOT_MASS_Y = 6.35;
+double ROBOT_MASS_H = 6.35 * 0.37;
+double ROBOT_RADIUS = 0.0794;
+
+void apply_wheel_force(const Eigen::Matrix<double, 4, 1> force, const Eigen::Matrix<double, 4, 1> speeds, Eigen::Matrix<double, 4, 1>& outputs) {
+    double CURRENT_PER_TORQUE = 1.0 / 25.1e-3;
+    double PHASE_RESISTANCE = 0.464;
+    for (int i = 0; i < 4; i++) {
+        double torque = force(i) * RobotModel::get().WheelRadius / 3.0;
+        double voltage = torque * CURRENT_PER_TORQUE * PHASE_RESISTANCE / 24.0;
+        double back_emf = (double) speeds(i) * RobotModel::get().SpeedToDutyCycle / 512.0;
+        outputs(i, 0) = 0.14 * voltage + back_emf * 0.7;
+    }
+}
+
+RobotController::RobotController(uint32_t dt_us)
+    : BodyUseILimit(true), BodyInputLimited(false),
+      BodyOutputLimited(false), dt(dt_us/1000000.0) {
+
+    // Body
+    BodyKp << 0, 0, 0.0; //1, 1.5, 1;
+    BodyKi << 0, 0, 0; //0.02, 0.02, 0.02;
+
+    BodyErrorSum << 0, 0, 0;
+    BodyILimit << 0.5, 0.5, 2;
+
+    BodyPrevTarget << 0, 0, 0;
+
+    // Wheel
+    // Gains should be the same across all wheels for now
+    WheelKp << 1.0, 1.0, 1.0, 1.0; //1.5
+
+    WheelPrevTarget << 0, 0, 0, 0;
+}
+
+void RobotController::calculateBody(Eigen::Matrix<double, numStates, 1> pv,
+                                    Eigen::Matrix<double, numStates, 1> sp,
+                                    Eigen::Matrix<double, numWheels, 1>& outputs) {
+    if (std::abs(sp(0)) > 6.0) {
+        sp(0) = sgn(sp(0)) * 6.0;
+    }
```

Comment:
This can be replaced with `std::clamp`.

Also, 6 seems _really really really really_ high for sideways motion, right?